### PR TITLE
Define Authenticator interface and a command implementation.

### PR DESCRIPTION
**Pull request by rgooch**
Keymasterd should be refactored to use this new Authenticator interface and instantiating a specific implementation based on the configuration. Existing authenticators should be changed to implement this interface and moved under lib/authenticate/*

### CODE REVIEW

File: lib/authenticate/command/impl.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		return nil, err
+	}
+	return &Authenticator{command, args, logger}, nil
+}
+
+func (au *Authenticator) authenticate(username, password string) bool {
+	args := []string{username}
+	args = append(args, au.args...)
+	cmd := exec.Command(au.command, args...)
+	cmd.Stdin = strings.NewReader(password + "\n")
```
Comment (cviecco) [inline lib/authenticate/command/impl.go]: Dont like passing "\n".... cant we just close stdin?

File: lib/authenticate/api.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package authenticate
+
+// Authenticator is an interface type that defines how to authenticate a user
+// with a username and password.
+type Authenticator interface {
+	// Authenticate will authenticate a user using the provided username and
+	// password. It returns true if the user is authenticated, else false (due
+	// to either invalid username or incorrect password).
+	Authenticate(username, password string) bool
```
Comment (cviecco) [inline lib/authenticate/api.go]: This inteface is problematic:
1. Passwords should be []byte
2. We should always be able to return errors, as the implementations might find errors such as network/other unknown conditions.

Also the name is problematic... I would call it passwordAuthenticate or similar

File: lib/authenticate/command/impl.go (hunk 1 above)
Comment (rgooch) [reply lib/authenticate/command/impl.go]: Done.

**Comment by rgooch** [comment]
PTAL.

**Comment by rgooch** [comment]
PTAL.

**Pull request merged**